process/dataProcessor.go

```diff
@@ -80,78 +83,128 @@ func (dp *dataProcessor) saveBlock(marshalledData []byte) error {
 
-	return nil
-}
+		log.Trace("do not commit block", "currentRound", metaRound, "firstCommitableRound", dp.firstCommitableBlock)
 
-func (dp *dataProcessor) putMetaNonce(nonce uint64, hash []byte) error {
-	nonceBytes := make([]byte, 8)
-	binary.LittleEndian.PutUint64(nonceBytes, nonce)
+		lastCheckpoint := &data.BlockCheckpoint{
+			LastRounds: map[uint32]uint64{
+				core.MetachainShardId: metaRound,
+			},
+		}
+		err = dp.outportBlocksPool.UpdateMetaState(lastCheckpoint)
+		if err != nil {
+			return err
+		}
 
-	return dp.outportBlocksPool.Put(nonceBytes, hash)
-}
+		return nil
+	}
 
-func (dp *dataProcessor) pushToBlocksPool(outportBlock *outport.OutportBlock) error {
-	blockHash := outportBlock.BlockData.HeaderHash
+	hyperOutportBlock, err := dp.dataAggregator.ProcessHyperBlock(metaOutportBlock)
+	if err != nil {
+		return err
+	}
 
-	header, err := dp.getHeader(outportBlock)
+	lastCheckpoint, err := dp.getLastRoundsData(hyperOutportBlock)
 	if err != nil {
 		return err
 	}
 
-	err = dp.outportBlocksPool.PutBlock(blockHash, outportBlock, header.GetRound())
+	err = dp.publisher.PublishHyperBlock(hyperOutportBlock)
 	if err != nil {
 		return err
 	}
 
-	return nil
+	return dp.outportBlocksPool.UpdateMetaState(lastCheckpoint)
 }
 
-func (dp *dataProcessor) getHeader(outportBlock *outport.OutportBlock) (data.HeaderHandler, error) {
-	blockCreator, err := dp.blockCreator.Get(core.HeaderType(outportBlock.BlockData.HeaderType))
-	if err != nil {
-		return nil, err
+// TODO: update to use latest data structures
+func (dp *dataProcessor) getLastRoundsData(hyperOutportBlock *hyperOutportBlocks.HyperOutportBlock) (*data.BlockCheckpoint, error) {
+	if hyperOutportBlock == nil {
+		return nil, ErrNilHyperOutportBlock
+	}
+	if hyperOutportBlock.MetaOutportBlock == nil {
+		return nil, fmt.Errorf("%w for metaOutportBlock", ErrNilHyperOutportBlock)
+	}
+	if hyperOutportBlock.MetaOutportBlock.BlockData == nil {
+		return nil, fmt.Errorf("%w for blockData", ErrNilHyperOutportBlock)
+	}
+	if hyperOutportBlock.MetaOutportBlock.BlockData.Header == nil {
+		return nil, fmt.Errorf("%w for blockData header", ErrNilHyperOutportBlock)
 	}
 
-	header, err := block.GetHeaderFromBytes(dp.marshaller, blockCreator, outportBlock.BlockData.HeaderBytes)
+	checkpoint := &data.BlockCheckpoint{
+		LastRounds: make(map[uint32]uint64),
+	}
+
+	metaBlock := hyperOutportBlock.MetaOutportBlock.BlockData.Header
+	checkpoint.LastRounds[core.MetachainShardId] = metaBlock.GetRound()
+
+	for _, outportBlockData := range hyperOutportBlock.NotarizedHeadersOutportData {
+		header := outportBlockData.OutportBlock.BlockData.Header
+		checkpoint.LastRounds[outportBlockData.OutportBlock.ShardID] = header.GetNonce()
+	}
+
+	return checkpoint, nil
+}
+
+func (dp *dataProcessor) handleShardOutportBlock(outportBlock *outport.OutportBlock) error {
+	shardOutportBlock, err := dp.outportBlockConverter.HandleShardOutportBlock(outportBlock)
 	if err != nil {
-		return nil, err
+		return err
 	}
```

Comment:
maybe a `fmt.Errorf("failed to convert outportBlock to shardOutportBlock: %w", err)`